File: 278-FirstBadVersion.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */
      
//Attention to overflow!!!
public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        if (n <= 0) return 0;
        if (n == 1) return 1;
        if (n == 2) return (isBadVersion(1) ? 1 : 2);
        int start = 0, end = n;
        while (end - start > 1){
            int mid = start + (end - start) / 2;
            if (isBadVersion(mid)){
                end = mid;
            }else{
                start = mid;
            }
        }
        return end;
    }
}

File: cpp/278-FirstBadVersion.cpp
// Forward declaration of isBadVersion API.
//The only scenario left is where isBadVersion(mid)⇒trueisBadVersion(mid)
//\Rightarrow trueisBadVersion(mid)⇒true. This tells us that midmidmid may
//or may not be the first bad version, but we can tell for sure that all
//versions after midmidmid can be discarded. Therefore we set right=midright = midright=mid
//as the new search space of interval [left,mid][left,mid][left,mid] (inclusive).
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int start = 1, end = n;
        while (start < end) {
            int mid = start + (end - start) / 2;
            if (isBadVersion(mid))
                end = mid;
            else
                start = mid + 1;
        }
        return start;
    }
};
